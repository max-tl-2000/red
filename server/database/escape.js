/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import mapValues from 'lodash/mapValues';
import stringify from 'json-stringify-safe';

import { isObject } from '../../common/helpers/type-of';

const doEscapeSingleQuotes = (value, processed) => {
  if (typeof value === 'string') return value.replace(/'/g, "''");

  const isArray = Array.isArray(value);
  const isObj = isObject(value);

  if (isArray || isObj) {
    if (processed.has(value)) {
      return value;
    }

    processed.add(value);
  }

  if (Array.isArray(value)) return value.map(entry => doEscapeSingleQuotes(entry, processed));
  if (value && typeof value === 'object') return mapValues(value, entry => doEscapeSingleQuotes(entry, processed));
  return value;
};

export const escapeSingleQuotes = value => doEscapeSingleQuotes(value, new WeakSet());

export const serializeAndEscapeSingleQuotes = value => {
  let hasCycles = false;

  const cycleDetector = () => {
    hasCycles = true;
    return 'Circular';
  };

  const serialized = stringify(escapeSingleQuotes(value), null, 0, cycleDetector);

  return { hasCycles, serialized };
};
