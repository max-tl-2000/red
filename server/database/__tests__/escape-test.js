/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import stringify from 'json-stringify-safe';

import { escapeSingleQuotes, serializeAndEscapeSingleQuotes } from '../escape.js';

describe('escape tests', () => {
  describe('when escaping single quotes in string', () => {
    it('should double the quotes', () => {
      expect(escapeSingleQuotes("that's it, it' enough")).toEqual("that''s it, it'' enough");
    });
  });
  describe('when escaping single quotes in non-string value', () => {
    it('should not alter the value', () => {
      expect(escapeSingleQuotes(42)).toEqual(42);
    });
  });
  describe('when escaping single quotes in an object', () => {
    it('should double the quotes in every string property value', () => {
      const obj = {
        a: "that's it",
        b: "who's there",
        c: 42,
      };
      const expected = {
        a: "that''s it",
        b: "who''s there",
        c: 42,
      };
      expect(escapeSingleQuotes(obj)).toEqual(expected);
    });
  });
  describe('when escaping single quotes in an array', () => {
    it('should double the quotes in every string element', () => {
      const array = [
        42,
        "it's time",
        {
          a: "that's it",
          b: "who's there",
          c: 42,
        },
      ];

      const expected = [
        42,
        "it''s time",
        {
          a: "that''s it",
          b: "who''s there",
          c: 42,
        },
      ];
      expect(escapeSingleQuotes(array)).toEqual(expected);
    });
  });

  describe('when escaping single quotes in an object cicular object', () => {
    it('should not throw', () => {
      const array = [
        42,
        "it's time",
        {
          a: "that's it",
          b: "who's there",
          c: 42,
        },
      ];

      array[2].arr = array;

      expect(stringify(escapeSingleQuotes(array), null, 2)).toMatchSnapshot();
    });
    describe('serializeAndEscapeSingleQuotes', () => {
      it('should detect cycles if any', () => {
        const array = [
          42,
          "it's time",
          {
            a: "that's it",
            b: "who's there",
            c: 42,
          },
        ];

        array[2].arr = array;

        const { serialized, hasCycles } = serializeAndEscapeSingleQuotes(array);

        expect(serialized).toMatchSnapshot();
        expect(hasCycles).toEqual(true);
      });

      it('should not detect cycles if none', () => {
        const array = [
          42,
          "it's time",
          {
            a: "that's it",
            b: "who's there",
            c: 42,
          },
        ];

        const { serialized, hasCycles } = serializeAndEscapeSingleQuotes(array);

        expect(serialized).toMatchSnapshot();
        expect(hasCycles).toEqual(false);
      });
    });
  });
});
