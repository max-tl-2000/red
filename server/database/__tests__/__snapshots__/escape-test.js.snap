// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`escape tests when escaping single quotes in an object cicular object serializeAndEscapeSingleQuotes should detect cycles if any 1`] = `"[42,\\"it''s time\\",{\\"a\\":\\"that''s it\\",\\"b\\":\\"who''s there\\",\\"c\\":42,\\"arr\\":[42,\\"it's time\\",{\\"a\\":\\"that's it\\",\\"b\\":\\"who's there\\",\\"c\\":42,\\"arr\\":\\"Circular\\"}]}]"`;

exports[`escape tests when escaping single quotes in an object cicular object serializeAndEscapeSingleQuotes should not detect cycles if none 1`] = `"[42,\\"it''s time\\",{\\"a\\":\\"that''s it\\",\\"b\\":\\"who''s there\\",\\"c\\":42}]"`;

exports[`escape tests when escaping single quotes in an object cicular object should not throw 1`] = `
"[
  42,
  \\"it''s time\\",
  {
    \\"a\\": \\"that''s it\\",
    \\"b\\": \\"who''s there\\",
    \\"c\\": 42,
    \\"arr\\": [
      42,
      \\"it's time\\",
      {
        \\"a\\": \\"that's it\\",
        \\"b\\": \\"who's there\\",
        \\"c\\": 42,
        \\"arr\\": \\"[Circular ~.2.arr]\\"
      }
    ]
  }
]"
`;
